man/man1/pmfind_check.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2020 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.TH PMFIND_CHECK 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmfind_check\f1 \- administration of Performance Co-Pilot discovery
.SH SYNOPSIS
.B $PCP_BINADM_DIR/pmfind_check
[\f3\-CNV?\f1]
[\f3\-l\f1 \f2logfile\f1]
.SH DESCRIPTION
This shell script is used to integrate the Performance Co-Pilot (see
.BR PCPIntro (1))
collector service discovery mechanisms with
.BR pmie (1)
and
.BR pmlogger (1)
service administration, such that multiple collector hosts can be
monitored from a single centralized host.
It is important to note that the
.B pmfind
service is tightly integrated with the
.B pmie
and
.B pmlogger
services, and these must also be enabled if the services are to
be started for each collector system that
.BR pmfind
discovers.
.PP
.B pmfind_check
is designed to be enabled via a service management daemon such as
.BR systemd (1)
on Linux.
It is run from a system timer (either by
.B systemd
or by
.B cron
on other systems).
Once per timer interval
.B pmfind
attempts discovery of PCP collector systems.
For each unique system found
.B pmfind_check
creates
.BR pmie_check
and
.BR pmlogger_check
control file entries, such that these latter process can manage one
.B pmie
and
.B pmlogger
service for each discovered collector system.
.PP
When run through
.B systemd
these processes will be launched immediately because
.B systemd
monitors the service control files for changes.
With
.B systemd
one may also initiate an immediate service discovery with
.B pmfind
and
.B pmfind_check
by modifying the
.IR $PCP_SYSCONF_DIR/pmfind
directory in any way.
.PP
When run from
.B cron
the next (also timer-based) invocation of
.B pmie_check
or
.B pmlogger_check
will start and manage the associated
.B pmie
and
.B pmlogger
processes.
.PP
The control file entries created by
.B pmfind_check
follow the convention of individual files for each collector
host.
The files are named using the (unique) source identifier that
.B pmfind
calculates.
This identifier is a hash calculated based on (non-optional)
context labels available from every collector host, and is the
same source identifier reported and used by
.BR pminfo (1)
and
.BR pmseries (1).
.SH OPTIONS
The available command line options are:
.TP 5
\fB\-C\fR, \fB\-\-containers\fR
This option causes
.B pmfind_check
to query each discovered collector system for running containers
as well, which will then be individually monitored.
.TP
\fB\-l\fR \fIfile\fR, \fB\-\-logfile\fR=\fIfile\fR
In order to ensure that mail is not unintentionally sent when these
scripts are run from
.BR cron (8)
diagnostics are always sent to log files.
By default, this file is
.B $PCP_LOG_DIR/pmfind/pmfind_check.log
but this can be changed using the
.B \-l
option.
If this log
.I file
already exists when the script starts, it will be renamed with a
.I .prev
suffix (overwriting any log file saved earlier) before diagnostics
are generated to the new log file.
.TP
\fB\-N\fR, \fB\-\-showme\fR
This option enables a ``show me'' mode, where the programs actions are
echoed, but not executed, in the style of ``make \-n''.
Using
.B \-N
in conjunction with
.B \-V
maximizes the diagnostic capabilities for debugging.
.TP
\fB\-V\fR, \fB\-\-verbose\fR
The output from the
.BR cron
execution of the scripts may be extended using the
.B \-V
option to the scripts which will enable verbose tracing of their activity.
By default the scripts generate no output unless some error or warning
condition is encountered.
Using
.B \-N
in conjunction with
.B \-V
maximizes the diagnostic capabilities for debugging.
.TP
\fB\-?\fR, \fB\-\-help\fR
Display usage message and exit.
.SH CONFIGURATION
The command line options passed to
.B pmie
and
.B
pmlogger
can be customized via the
.IR $PCP_SYSCONFIG_DIR/pmfind
configuration file.
.PP
Refer to the
.B pmie_check
and
.B pmlogger_check
manual pages for details of the control file format generated by
.BR pmfind_check .
.PP
By default, each collector system
.B pmie
logfile will be created in
.BI $PCP_LOG_DIR/pmie/ source
and
.B pmlogger
archives will be created below
.BI $PCP_ARCHIVE_DIR/ source\fR.
This is consistent with the way
.BR pmie (1)
and
.BR pmlogger (1)
maintain their activity logs and archive files.
.PP
Typical
.BR crontab (5)
entries for periodic execution of
.B pmfind_check
are given in
.BR $PCP_SYSCONF_DIR/pmfind/crontab
(unless installed by default in
.IR /etc/cron.d
already)
and shown below.
.PP
.nf
.ft CR
# every minute, discover remote PCP services
*       *       *       *       *       $PCP_BINADM_DIR/pmfind_check
.ft 1
.fi
When using
.BR systemd (1)
on Linux,
no
.B crontab
entries are needed as the timer and path mechanisms provided by
.B systemd
are used instead.
.SH FILES
.TP 5
.I $PCP_PMIECONTROL_PATH.d
PCP inference engine control files for each discovered metric source,
typically one per host or container.
.br
.BR Warning :
this files herein must not be writable by any user other than root.
.TP
.I $PCP_PMLOGGERCONTROL_PATH.d
PCP logger control files for each discovered metric source,
typically one per host or container.
.br
.BR Warning :
this files herein must not be writable by any user other than root.
.TP
.I $PCP_SYSCONFIG_DIR/pmfind
default
.B pmie
and
.B pmlogger
command line arguments, and other parameters passed to
.BR pmfind .
.TP
.I $PCP_SYSCONF_DIR/pmfind/crontab
sample crontab for automated script execution by root -
exists only if the platform does not support the
.I /etc/cron.d
mechanism.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR pmcd (1),
.BR pmfind (1),
.BR pmie (1),
.BR pmie_check (1),
.BR pminfo (1),
.BR pmlogger (1),
.BR pmlogger_check (1),
.BR pmseries (1),
.BR systemd (1),
.BR crontab (5),
and
.BR cron (8).

.\" control lines for scripts/man-spell
.\" +ok+ prev {from .prev suffix} \* {dopey ispell}


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmfind/pmfind_check.sh
#! /bin/sh
#
# Copyright (c) 2020 Red Hat.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#
# Administrative script to discover remote PCP services and configure
# pmie and pmlogger control files for each as required.
#

# get standard environment
. $PCP_DIR/etc/pcp.env
. $PCP_SHARE_DIR/lib/rc-proc.sh
. $PCP_SHARE_DIR/lib/utilproc.sh

PMFIND=${PMFIND:-"$PCP_BIN_DIR/pmfind"}

PMFIND_ARGS=${PMFIND_ARGS:-'-s pmcd -r -S -q'}
CTL_ARGS=''

# error messages should go to stderr, not the GUI notifiers
unset PCP_STDERR

tmp=`mktemp -d "$PCP_TMPFILE_DIR/pmfind_check.XXXXXXXXX"` || exit 1
status=0
prog=`basename $0`
PROGLOG=$PCP_LOG_DIR/pmfind/$prog.log
USE_SYSLOG=true

_cleanup()
{
    $USE_SYSLOG && [ $status -ne 0 ] && \
    $PCP_SYSLOG_PROG -p daemon.error "$prog failed - see $PROGLOG"
    if $SHOWME
    then
	:
    elif [ "$PROGLOG" != "/dev/tty" ]
    then
	[ -s "$PROGLOG" ] || rm -f "$PROGLOG"
    fi
    rm -rf $tmp
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# option parsing
CONTAINERS=false
SHOWME=false
VERBOSE=false
VERY_VERBOSE=false

echo > $tmp/usage
cat >> $tmp/usage << EOF
Options:
  -C,--containers         probe each discovered host for pmcd containers
  -l=FILE,--logfile=FILE  send important diagnostic messages to FILE
  -N,--showme             perform a dry run, showing what would be done
  -V,--verbose            increase diagnostic verbosity
  --help
EOF

ARGS=`pmgetopt --progname=$prog --config=$tmp/usage -- "$@"`
[ $? != 0 ] && exit 1

eval set -- "$ARGS"
while [ $# -gt 0 ]
do
    case "$1"
    in
        -C)     CONTAINERS=true
		;;
	-l)	PROGLOG="$2"
		USE_SYSLOG=false
		shift
		;;
	-N)	SHOWME=true
		USE_SYSLOG=false
		CP="echo + cp"
		CTL_ARGS="$CTL_ARGS -N"
		;;
	-V)	if $VERBOSE
		then
		    VERY_VERBOSE=true
		else
		    VERBOSE=true
		fi
		CTL_ARGS="$CTL_ARGS -V"
		;;
	--)	shift
		break
		;;
	-\?)	pmgetopt --usage --progname=$prog --config=$tmp/usage
		status=1
		exit
		;;
    esac
    shift
done

if [ $# -ne 0 ]
then
    pmgetopt --usage --progname=$prog --config=$tmp/usage
    status=1
    exit
fi

if $SHOWME
then
    :
else
    # Salt away previous log, if any ...
    _save_prev_file "$PROGLOG"

    # After argument checking, everything must be logged to ensure no mail is
    # accidentally sent from cron.  Close stdout and stderr, then open stdout
    # as our logfile and redirect stderr there too.
    #
    # Exception ($SHOWME, above) is for -N where we want to see the output.
    #
    exec 1>"$PROGLOG" 2>&1
fi

_error()
{
    echo "$prog: [$SHA1]"
    echo "Error: $@"
    echo "... automated service discovery for host \"$host\" unchanged"
    touch $tmp/err
}

_debug()
{
    if $VERY_VERBOSE
    then
	echo "$prog [$SHA1]"
	echo "Debug: $@"
    fi
}

$CONTAINERS && PMFIND_ARGS="$PMFIND_ARGS -C"
$PMFIND $PMFIND_ARGS > "$tmp/out" 2> "$tmp/err"
# exit status from pmfind(1) is not a good indicator of success,
# no output => nothing to do, status == 0 => work to be done
#
[ -s "$tmp/out" ] || exit
[ $? -eq 0 ] && rm -f "$tmp/err"

pmiectl -V status >$tmp/pmie.status 2>&1
pmlogctl -V status >$tmp/pmlogger.status 2>&1

cat "$tmp/out" | while read SHA1 host
do
    $VERBOSE && echo "Discovered host $host [$SHA1]"

    if grep " $SHA1\$" $tmp/pmie.status >/dev/null
    then
	_debug "pmie already setup for host \"$host\""
    else
	# no pmie has instance id $SHA1, set it up
	#
	if pmiectl $CTL_ARGS -i "$SHA1" cond-create $host
	then
	    _debug "pmie setup for host \"$host\""
	else
	    _debug "pmie setup failed for host \"$host\""
	    touch $tmp/err
	fi
    fi

    if grep " $SHA1\$" $tmp/pmlogger.status >/dev/null
    then
	_debug "pmlogger already setup for host \"$host\""
    else
	# no pmlogger has instance id $SHA1, set it up
	#
	if pmlogctl $CTL_ARGS -i "$SHA1" cond-create $host
	then
	    _debug "pmlogger setup for host \"$host\""
	else
	    _debug "pmlogger setup failed for host \"$host\""
	    touch $tmp/err
	fi
    fi
done

[ -f $tmp/err ] && status=1
exit
